File: frontend/src/types/structural-element.ts
// Mirrors app/schemas/structural_element.py exactly.
import type { ElementType } from "@/lib/constants";

export interface StructuralElement {
  id: number;
  element_name: string;
  element_type: ElementType;
  material: string | null;
  floor_id: number;
  room_id: number | null;
  dim_width: number | null;
  dim_height: number | null;
  dim_depth: number | null;
}

export interface StructuralElementCreateInput {
  element_name: string;
  element_type: ElementType;
  material?: string | null;
  floor_id: number;
  room_id?: number | null;
  dim_width?: number | null;
  dim_height?: number | null;
  dim_depth?: number | null;
}

// StructuralElementUpdate has no floor_id/room_id — reassignment isn't supported by this API.
export interface StructuralElementUpdateInput {
  element_name?: string;
  element_type?: ElementType;
  material?: string | null;
  dim_width?: number | null;
  dim_height?: number | null;
  dim_depth?: number | null;
}

File: backend/app/schemas/structural_element.py
from pydantic import BaseModel, ConfigDict

from app.domain.enums import ElementType


class StructuralElementCreate(BaseModel):
    element_name: str
    element_type: ElementType
    material: str | None = None
    floor_id: int
    room_id: int | None = None
    dim_width: float | None = None
    dim_height: float | None = None
    dim_depth: float | None = None


class StructuralElementUpdate(BaseModel):
    element_name: str | None = None
    element_type: ElementType | None = None
    material: str | None = None
    dim_width: float | None = None
    dim_height: float | None = None
    dim_depth: float | None = None


class StructuralElementResponse(BaseModel):
    id: int
    element_name: str
    element_type: ElementType
    material: str | None
    floor_id: int
    room_id: int | None
    dim_width: float | None
    dim_height: float | None
    dim_depth: float | None

    model_config = ConfigDict(from_attributes=True)
